newlib is missing hardware floating support

How can I use this newly packaged `newlib` with hardware floating point support?

I have the following code example:

```nix
# The commit right before your removed gcc-arm-embedded: c327df544ae0875ce71b0a82ae7a650a8e34a268
with import (fetchTarball "https://github.com/nixos/nixpkgs/archive/aa4707fc3449192b961bfd2dfc16b9042df0d7f3.tar.gz") {};
rec {
  # This is a small c file which uses `log` from libm
  file = writeTextFile {
    name = "file.c";
    text = ''
      #include <math.h>
      int _exit(int i){};  # missing symbol during build, this can be ignored.
      int main()
      {
        float v = log(0.f);
        return 0;
      }
    '';
  };

  # my build command ligne, note the -mfloat-abi=hard
  cmd = ''arm-none-eabi-gcc -mfloat-abi=hard ${file} -o file -lm -o $out'';

  # a generic buider
  build = compiler : runCommand "foo" {
    allowSubstitutes = false;
    nativeBuildInputs = [compiler];
  }
  ''arm-none-eabi-gcc -mfloat-abi=hard ${file} -o file -lm -o $out'';

  # that one success
  success = build gcc-arm-embedded;
  # that one fails, with /nix/store/z6pvjan22wqxcg787kyw89jl64dhh2ks-arm-none-eabi-binutils-2.30/bin/arm-none-eabi-ld: error: /build/ccDue1Qh.o uses VFP register arguments, /nix/store/nwf9pkcgjlcympb2yfqwx2svim6w6sa9-foo does not
  failure = build pkgsCross.arm-embedded.buildPackages.gcc;
}
```

Please not the `-mfloat-abi=hard` flag. Apparently, by default, `newlib` is built without hardware floating point support. See https://sourceware.org/newlib/faq.html#q4 which discuss the build flag `--newlib-hw-fp`:

> The `newlib-hw-fp` option causes the library to use a set of math routines that use floating-point algorithms. By default, the newlib library math routines use integer-based algorithms to perform the desired functionality. For a platform that has fast floating-point support, the floating-point algorithms may be faster and smaller.

I'v tried to override `newlib` or `newlibCross` without success with an overlay:

```
 newlibCross = super.newlibCross.overrideAttrs (old : {
    configureFlags = old.configureFlags ++ ["--enable-newlib-hw-fp"];
  });
```

And I've tried to import my cross system using `import <nixpkgs> { crossSystem = { ...; float = "hard" }; }` without success either.

_Originally posted by @guibou in https://github.com/NixOS/nixpkgs/issues/47533#issuecomment-441060860_

@Ericson2314 

About `eabihf`, what do you mean? Based on the example for `arm-none-eabi`, I tried a few variation of:

```
import <nixpkgs> {
  crossSystem = {
    config = XXX;
    libc = "newlib";
  };
};
```

where `XXX` was:

- `arm-none-eabihf`: nix error `Target specification with 3 components is ambiguous`
- `arm-none-gnueabihf`: same
- `arm-unknown-none-eabihf`: nix error `Unknown ABI: eabihf`
- `arm-unknown-none-gnueabihf`: build error:

```
...
checking target system type... Invalid configuration `arm-unknown-none-gnueabihf': machine `arm-unknown-none' not recognized
configure: error: /nix/store/r47p5pzx52m3n34vdgqpk5rvqgm0m24m-bash-4.4-p23/bin/bash ./config.sub arm-unknown-none-gnueabihf failed
builder for '/nix/store/317r53lkbpdfmi4v39gc4zi263py1ls8-arm-unknown-none-gnueabihf-binutils-2.30.drv' failed with exit code 1
```

--- a/lib/systems/examples.nix
+++ b/lib/systems/examples.nix
@@ -110,6 +110,10 @@ rec {
     config = "arm-none-eabi";
     libc = "newlib";
   };
+  armhf-embedded = {
+    config = "arm-none-eabihf";
+    libc = "newlib";
+  };
 
   aarch64-embedded = {
     config = "aarch64-none-elf";

--- a/lib/systems/parse.nix
+++ b/lib/systems/parse.nix
@@ -209,8 +209,15 @@ rec {
   abis = setTypes types.openAbi {
     cygnus       = {};
     msvc         = {};
-    eabi         = {};
-    elf         = {};
+
+    # Note: eabi is specific to ARM and PowerPC.
+    # On PowerPC, this corresponds to PPCEABI.
+    # On ARM, this corresponds to ARMEABI.
+    eabi         = { float = "soft"; };
+    eabihf       = { float = "hard"; };
+
+    # Other architectures should use ELF in embedded situations.
+    elf          = {};
 
     androideabi  = {};
     android      = {
@@ -272,10 +279,8 @@ rec {
     "2" = # We only do 2-part hacks for things Nix already supports
       if elemAt l 1 == "cygwin"
         then { cpu = elemAt l 0;                      kernel = "windows";  abi = "cygnus";   }
-      else if (elemAt l 1 == "eabi")
-        then { cpu = elemAt l 0; vendor = "none"; kernel = "none"; abi = elemAt l 1; }
-      else if (elemAt l 1 == "elf")
-        then { cpu = elemAt l 0; vendor = "none"; kernel = "none"; abi = elemAt l 1; }
+      else if (elemAt l 1) == "elf"
+      then { cpu = elemAt l 0; vendor = "unknown";    kernel = "none";     abi = elemAt l 1; }
       else   { cpu = elemAt l 0;                      kernel = elemAt l 1;                   };
     "3" = # Awkwards hacks, beware!
       if elemAt l 1 == "apple"
@@ -286,10 +291,8 @@ rec {
         then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "windows";  abi = "gnu"; }
       else if hasPrefix "netbsd" (elemAt l 2)
         then { cpu = elemAt l 0; vendor = elemAt l 1;    kernel = elemAt l 2;                }
-      else if (elemAt l 2 == "eabi")
-        then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "none"; abi = elemAt l 2; }
-      else if (elemAt l 2 == "elf")
-        then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "none"; abi = elemAt l 2; }
+      else if (elem (elemAt l 2) ["eabi" "eabihf" "elf"])
+        then { cpu = elemAt l 0; vendor = "unknown"; kernel = elemAt l 1; abi = elemAt l 2; }
       else throw "Target specification with 3 components is ambiguous";
     "4" =    { cpu = elemAt l 0; vendor = elemAt l 1; kernel = elemAt l 2; abi = elemAt l 3; };
   }.${toString (length l)}

--- a/pkgs/development/misc/qmk_firmware/default.nix
+++ b/pkgs/development/misc/qmk_firmware/default.nix
@@ -1,6 +1,6 @@
 { stdenv, fetchFromGitHub
 , avrgcc, avrbinutils
-, gcc-arm-embedded, binutils-arm-embedded
+, gcc-arm-embedded, gcc-armhf-embedded
 , teensy-loader-cli, dfu-programmer, dfu-util }:
 
 let version = "0.6.144";
@@ -14,12 +14,23 @@ in stdenv.mkDerivation {
     sha256 = "0m71f9w32ksqjkrwhqwhr74q5v3pr38bihjyb9ks0k5id0inhrjn";
     fetchSubmodules = true;
   };
+  postPatch = ''
+    substituteInPlace tmk_core/arm_atsam.mk \
+      --replace arm-none-eabi arm-none-eabihf
+    rm keyboards/handwired/frenchdev/rules.mk keyboards/dk60/rules.mk
+  '';
   buildFlags = "all:default";
+  doCheck = true;
+  checkTarget = "test:all";
+  installPhase = ''
+    mkdir $out
+  '';
   NIX_CFLAGS_COMPILE = "-Wno-error";
   nativeBuildInputs = [
     avrgcc
     avrbinutils
     gcc-arm-embedded
+    gcc-armhf-embedded
     teensy-loader-cli
     dfu-programmer
     dfu-util

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -23027,7 +23027,7 @@ with pkgs;
     avrgcc = pkgsCross.avr.buildPackages.gcc;
     avrbinutils = pkgsCross.avr.buildPackages.binutils;
     gcc-arm-embedded = pkgsCross.arm-embedded.buildPackages.gcc;
-    binutils-arm-embedded = pkgsCross.arm-embedded.buildPackages.binutils;
+    gcc-armhf-embedded = pkgsCross.armhf-embedded.buildPackages.gcc;
   };
 
   newlib = callPackage ../development/misc/newlib { };

